### PR TITLE
Backport chacha-poly1305, prepare for V2 P2P Encrypted Messaging 

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
This backports #15512 #15519 and #15649. This prepares us to be able to backport the v2 encrypted messaging system once the implementation is finalized in bitcoin. We could also not merge this until bitcoin is done implementing v2 messaging system, but since I would want a lot of manual testing on the v2 messaging system (and it might have a lot of merge conflicts) I think it makes sense to do this as a seperate backporting PR. 

The PR to implement the V2 messaging system will likely end up being https://github.com/bitcoin/bitcoin/pull/14032

### CODE REVIEW

**Comment by UdjinM6** [comment]
You need #11196 to fix windows builds.

In general, IMO PRs like 14032 can take years to be merged into Bitcoin Core, so I don't see a reason to rush with backporting chacha ones. On the other hand, these changes seems to be isolated enough and shouldn't cause merge conflicts in the future, so it's probably ok to merge them anyway :)

**Comment by PastaPastaPasta** [comment]
On the bright side 14032 has been open 11 months already and the related PRs have been getting a lot of attention

**Comment by UdjinM6** [comment]
I have an email notification about commit https://github.com/dashpay/dash/commit/60ba106ccf7ec59a5e93d5b06119c166f90fcac3 pushed to this PR but Github isn't showing it here for some reason. Might want to force-push it.

**Comment by PastaPastaPasta** [comment]
force pushed

File: src/bench/chacha20.cpp [hunk 1]
```diff
@@ -41,6 +41,10 @@ static void CHACHA20_1MB(benchmark::State& state)
...
-BENCHMARK(CHACHA20_64BYTES, 500000);
-BENCHMARK(CHACHA20_256BYTES, 250000);
-BENCHMARK(CHACHA20_1MB, 340);
+//TODO add back below once benchmarking backports are done
+//BENCHMARK(CHACHA20_64BYTES, 500000);
+//BENCHMARK(CHACHA20_256BYTES, 250000);
+//BENCHMARK(CHACHA20_1MB, 340);
+BENCHMARK(CHACHA20_64BYTES);
+BENCHMARK(CHACHA20_256BYTES);
+BENCHMARK(CHACHA20_1MB);
```
Comment (codablock) [inline src/bench/chacha20.cpp]: Don't let Nathan see this :P (missing new-line)

Comment (nmarley) [reply src/bench/chacha20.cpp]: :D I actually did notice this yesterday. It sticks out like a sore thumb. 👍

Comment (PastaPastaPasta) [reply src/bench/chacha20.cpp]: resolved and then squashed

**Pull request merged**